cmd/istioctl/main.go

```diff
@@ -314,8 +335,12 @@ func init() {
 
 	postCmd.PersistentFlags().StringVarP(&file, "file", "f", "",
 		"Input file with the content of the configuration objects (if not set, command reads from the standard input)")
+	postCmd.PersistentFlags().StringVar(&kubectlcommand, "kubectl", "",
+		"The path to kubectl command for uploading mixer config. If not set, kubectl will be found from $PATH.")
 	putCmd.PersistentFlags().AddFlag(postCmd.PersistentFlags().Lookup("file"))
+	putCmd.PersistentFlags().AddFlag(postCmd.PersistentFlags().Lookup("kubectl"))
 	deleteCmd.PersistentFlags().AddFlag(postCmd.PersistentFlags().Lookup("file"))
+	deleteCmd.PersistentFlags().AddFlag(postCmd.PersistentFlags().Lookup("kubectl"))
```

Comment:
Since we already take KUBECONFIG as a flag, wouldn't it be better to use dynamic REST mapper to send Mixer's CRDs? What's the obstacle?

Reply:
I argue that we stop doing that (and trying to access some hidden config map, if we are still doing it) even when pushing proxy CRDs.
We don't want compile time dependencies on k8s. We just need the `kubectl` binary from somewhere.

`Istioctl` should
1. Support start CRUD verbs
2. Know the proxy types
3. All other types are considered "Other istio types"
4. Automagically add namespace / apiGroup to the yaml files on the way out.
5. Potentially call some external validator until webhook works without alpha

All this, like I said yesterday *can* be done with a shell script. We just want a little bit more control, and shell scripts aren't great for that.

If there is a need to interact directly with the API, beyond what kubectl can do, we can have compile time deps then.

Otherwise we have to replicate all the cli work done by `kubectl`, things like `kubectl apply`

Reply:
The only reason to use API directly is when using in non-k8s env. Or `openshift` command line. Even then, we can probably bundle `kubectl`

Reply:
1. `kubectl apply` does not work well for CRDs. So we have an opportunity to do better if we choose to. Using a binary dependency complicates installation, since it's a surprise that installing istioctl requires installing kubectl.
2. The hidden configmap is terrible, but it's getting better. It's only used by kube-inject to  get mesh addresses. It'll go away with in-cluster initializer.

Reply:
+1 to "Using a binary dependency complicates installation, since it's a surprise that installing istioctl requires installing kubectl.".

Reply:
```The only reason to use API directly is when using in non-k8s env. Or openshift command line. Even then, we can probably bundle kubectl```
Not bundling kubectl is the whole point! why would you force end users in cloud foundry to use kubectl to update routing rules for their CF app? If you put yourself in a UX person or a PM's shoe, you will notice that this notion is not only ugly but also feels "forced" (indirectly forcing kubernetes in end users who are not using it). That gets into political realm quickly.

Reply:
you can't use istioctl without a .kube/config and you would get that because you have kubectl
so I think it's safe/practical to assume kubectl is there - or - we need to somehow work without kube config (but we are talking to a k8s api server so why not)

Reply:
in 0.2 (though I am curious how does the consul setup configures/accesses the api server)

Reply:
removed, now using k8s.io/client-go libraries.